app/vmselect/promql/exec.go

```diff
@@ -13,10 +13,81 @@ import (
 	"github.com/VictoriaMetrics/VictoriaMetrics/lib/logger"
 	"github.com/VictoriaMetrics/metrics"
 	"github.com/VictoriaMetrics/metricsql"
+
+	"github.com/google/uuid"
 )
 
 var logSlowQueryDuration = flag.Duration("search.logSlowQueryDuration", 5*time.Second, "Log queries with execution time exceeding this value. Zero disables slow query logging")
 
+type Query struct {
+	q       string
+	ec      *EvalConfig
+	stopCh  chan error
+	startAt time.Time
+}
+
+type queriesMap struct {
+	mu sync.Mutex
+	m  map[string]Query
+	c  int64
+}
+
+func (rqm *queriesMap) Init() {
+	rqm.m = make(map[string]Query)
+}
+
+func (rqm *queriesMap) Add(q Query) string {
+	rqm.mu.Lock()
+	c := uuid.New().String()
+	rqm.m[c] = q
+	rqm.mu.Unlock()
+
+	return c
+}
+
+func (rqm *queriesMap) Delete(c string) {
+	rqm.mu.Lock()
+	delete(rqm.m, c)
+	rqm.mu.Unlock()
+}
+
+func (rqm *queriesMap) Kill(c string) {
+	rqm.mu.Lock()
+	// do some thing to kill the request
+	rqm.mu.Unlock()
+}
+
+var runningQueries queriesMap
+
+func init() {
+	runningQueries.Init()
+}
+
+func GetAllRunningQueries() map[string]map[string]string {
+	all := make(map[string]map[string]string)
+	runningQueries.mu.Lock()
+	for c, query := range runningQueries.m {
+		m := make(map[string]string)
+		m["query"] = query.q
+		m["cost"] = time.Since(query.startAt).String()
+		all[c] = m
+	}
+	runningQueries.mu.Unlock()
+
+	return all
+}
+
+func CancelRunningQuery(c string) error {
+	runningQueries.mu.Lock()
+	defer runningQueries.mu.Unlock()
+	if query, ok := runningQueries.m[c]; ok {
+		query.stopCh <- fmt.Errorf("cancel query manully")
```

Comment:
It looks like the `stopCh` handler isn't implemented in the query execution code. So I'd propose removing all the code related to query canceling from this pull request and implement it in the follow-up pull request.

Reply:
I save all running queries at `runningQueries `, and every query instance has a channel of type error:
```
type Query struct {
	...
	stopCh  chan error
}
```
and when the query is running, I listen this channel see if any error occurs, that's why i don't return in `exec` function at this review:
> [It looks like `return` is missing after such lines](https://github.com/VictoriaMetrics/VictoriaMetrics/pull/575#discussion_r443133150)

and why I use `go exec()` at this review:
> [It looks like it is better to use synchronous code here, i.e. calling `exec()` directly instead of `go exec()`, since `go exec()` complicates the code and makes it more fragile. For instance, the goroutine may continue executing or may block indefinitely on `stopCh` or `resultCh`. This may result in unbound CPU usage or goroutine leak.](https://github.com/VictoriaMetrics/VictoriaMetrics/pull/575#discussion_r443132973)

I think canceling is an error(made manually), so that's why I send a error to query instance's `stopChan` when I send a canceling request, then the execution receives the error and return, the go routine of `exec` will be stoped, the reason will be logged like this.
![image](https://user-images.githubusercontent.com/13900390/85215637-474afe80-b3ae-11ea-8945-8eb286720dfa.png)